test/spec/Feature/Query/UpdateSpec.hs

```diff
@@ -535,11 +541,137 @@ spec = do
+        ]|]
+      `shouldMutateInto`
+      [json|[
+        { "id": 1, "name": "item-1", "observation": null }
+      , { "id": 2, "name": "item-2 - 2nd", "observation": "Damaged item" }
+      , { "id": 3, "name": "item-3 - 3rd", "observation": null }
+      ]|]
+
+    it "updates with limit and offset taking only the first item in the json array body" $ do
+      baseTable "bulk_update_items" "id" tblDataBeforeBulk
+      `mutatesWith`
+      requestMutation methodPatch "/bulk_update_items?limit=2&offset=1&order=id"
+        [("Prefer", "params=multiple-objects")]
+        [json|[
+          { "name": "item", "observation": "Lost item" }
+        , { "name": "item-2 - 2nd", "observation": "Damaged item" }
+        , { "name": "item-3 - 3rd", "observation": null }
+        ]|]
+      `shouldMutateInto`
+      [json|[
+        { "id": 1, "name": "item-1", "observation": null }
+      , { "id": 2, "name": "item", "observation": "Lost item" }
+      , { "id": 3, "name": "item", "observation": "Lost item" }
+      ]|]
+
+    it "does a full table update without the header for bulk updates, taking only the first item in the json array body" $ do
+      baseTable "bulk_update_items" "id" tblDataBeforeBulk
+      `mutatesWith`
+      requestMutation methodPatch "/bulk_update_items" mempty
+        [json|[
+          { "name": "item", "observation": "Damaged item" }
+        , { "name": "item-3 - 3rd", "observation": null }
+        ]|]
+      `shouldMutateInto`
+      [json|[
+        { "id": 1, "name": "item", "observation": "Damaged item" }
+      , { "id": 2, "name": "item", "observation": "Damaged item" }
+      , { "id": 3, "name": "item", "observation": "Damaged item" }
+      ]|]
+
+    it "rejects a json array that isn't exclusively composed of objects" $
+      request methodPatch "/bulk_update_items"
+          [("Prefer", "tx=commit"), ("Prefer", "params=multiple-objects")]
+          [json|[
+            { "id": 1, "name": "Item 1", "observation": null }
+          , 2
+          , "Item 2"
+          , { "id": 3, "name": "Item 3" }
+          ]|]
+        `shouldRespondWith`
+          [json| {"message":"All object keys must match","code":"PGRST102","hint":null,"details":null} |]
+          { matchStatus  = 400
+          , matchHeaders = [matchContentTypeJson]
+          }
+
+    it "rejects a json array that has objects with different keys" $
+      request methodPatch "/bulk_update_items"
+          [("Prefer", "tx=commit"), ("Prefer", "params=multiple-objects")]
+          [json|[
+            { "id": 1, "name": "Item 1" }
+          , { "id": 2, "name": "Item 2" }
+          , { "id": 3, "name": "Item 3", "observation": "New item" }
+          ]|]
+        `shouldRespondWith`
+          [json| {"message":"All object keys must match","code":"PGRST102","hint":null,"details":null} |]
+          { matchStatus  = 400
+          , matchHeaders = [matchContentTypeJson]
+          }
```

Comment:
Hm, this is definitely undesirable. Users will want to update different columns for each array element. I think we'll need https://github.com/PostgREST/postgrest/pull/2672 for solving this.

Reply:
Maybe it can be solved without #2672. So assume you specify `columns` here or another test. You could do:
```sql
UPDATE bulk_update_items SET observation = COALESCE(pgrst_body.observation, observation) WHERE ..
```
 Meaning that a value will be updated to the new value in the body or to its old value.

This will allow updating rows with a changing number of columns. WDYT?

Reply:
After implementing this I noticed that setting a value to null will fail, e.g.:

```http
/bulk_update_items?columns=id,name,observation
```
```json
[
            { "id": 1, "name": "Item 1" }
          , { "id": 3, "name": "Item 3", "observation": null }
]
```

The item with `id=3` will keep its old `observation` and won't be set to `null`. Is there a conditional that could be used here, to only apply `COALESCE` in these cases? Although, I see it difficult due to the fact that `columns` does not check the body, right?

Reply:
Right. Wolfgang suggested a way to do this with the `?` operator on `jsonb` [here](https://github.com/PostgREST/postgrest/issues/1567#issuecomment-739478169).

On https://github.com/PostgREST/postgrest/pull/2672 I had the same problem as well, but I decided to go for the alternative of using the `||` operator on `jsonb`. 

Maybe those approaches could help.

Reply:
I ended up using `?` instead. Although it works now, I know that it isn't too performant. But when i used `||` , I got this query:

```sql
UPDATE "test"."bulk_update_items"
SET "id" = "pgrst_body"."id",
    "name" = "pgrst_body"."name",
    "observation" = "pgrst_body"."observation"
FROM
    (SELECT $1::jsonb AS json_data) pgrst_payload
    CROSS JOIN LATERAL
        (SELECT CASE WHEN jsonb_typeof(pgrst_payload.json_data) = 'array' THEN pgrst_payload.json_data ELSE jsonb_build_array(pgrst_payload.json_data) END AS val) pgrst_uniform_json
    CROSS JOIN LATERAL
        (SELECT jsonb_agg(jsonb_build_object('id', "test"."bulk_update_items"."id",'name', "test"."bulk_update_items"."name",'observation', "test"."bulk_update_items"."observation") || elem) AS val
         FROM jsonb_array_elements(pgrst_uniform_json.val) elem) pgrst_json_old_vals
    CROSS JOIN LATERAL
        (SELECT *
         FROM jsonb_to_recordset(pgrst_json_old_vals.val) AS _("id" integer, "name" text, "observation" text)) pgrst_body
WHERE "test"."bulk_update_items"."id" = "pgrst_body"."id" RETURNING 1;
```

Which gives an error when accessing `test.bulk_update_items` inside the `pgrst_json_old_vals` subquery due to the `CROSS JOIN`. Couldn't find a way to make it work that way, maybe I'm missing something.